tools/indexer/main.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2018 Microsoft Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+type packageSet map[string]bool
+
+// adds any missing SDK packages to godoc.org
+func main() {
+	dir := ""
+	if len(os.Args) > 1 {
+		// assume second arg is source dir
+		dir = os.Args[1]
+	} else {
+		// if no args specified assume we're running from the source dir
+		// and calculate the relative path to the services directory.
+		var err error
+		dir, err = filepath.Abs("../../services")
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	pkgs, err := getPackagesForIndexing(dir)
+	if err != nil {
+		panic(err)
+	}
+
+	indexedPkgs, err := getIndexedPackages()
+	if err != nil {
+		panic(err)
+	}
+
+	// for each package in pkgs, check if it's already been
+	// indexed.  if it hasn't been indexed then do so
+
+	for pkg := range pkgs {
+		if _, already := indexedPkgs[pkg]; already {
+			pkgs[pkg] = true
+			continue
+		}
+
+		fmt.Printf("indexing %s...", pkg)
+		resp, err := http.DefaultClient.Get(fmt.Sprintf("https://godoc.org/%s", pkg))
+		if err != nil {
+			panic(err)
+		}
+
+		resp.Body.Close()
+		if resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusNotFound {
+			// a 404 means that the package exists locally but not yet in github
+			fmt.Printf("completed with status '%s'\n", resp.Status)
+			pkgs[pkg] = true
+		} else {
+			fmt.Printf("FAILED with status '%s'\n", resp.Status)
+		}
+
+		// sleep a bit between indexing
+		time.Sleep(60 * time.Second)
+	}
+
+	// check if any packages failed to index
+	failed := false
+	for _, v := range pkgs {
+		if !v {
+			failed = true
+			break
+		}
+	}
+
+	if failed {
+		fmt.Println("not all packages were indexed")
+		os.Exit(1)
+	}
+	fmt.Println("successfully indexed all packages")
+}
+
+// returns the set of packages that have already been indexed
+func getIndexedPackages() (packageSet, error) {
```

Comment:
I presume this gets the href from the html page ? seems ok given that the format is "/github.com/Azure/azure-sdk-for-go/services/advisor/mgmt/2017-03-31/advisor" ?

If it works i think its ok but was wondering if it would be easier to just have a regex match on the pattern with a capture group for the actual pacakge name ( something like  this [https://regex101.com/r/r45IYk/1](https://regex101.com/r/r45IYk/1) ( needs the non greedy option set )

Reply:
Fixed.